### PR TITLE
Added Observer and Subscription for receiving push-based notifications

### PR DESCRIPTION
**Pull request by rozza**
JAVA-1805

[Review on Reviewable](https://reviewable.io/reviews/rozza/mongo-java-driver/94)

### CODE REVIEW

File: driver-async/src/main/com/mongodb/async/client/SubscriptionHelpers.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+
+/**
+ * Subscription helpers.
+ * <p/>
+ * <p>Allows async methods to be converted into event-based style observables.</p>
+ *
+ * @since 3.1
+ */
+public final class SubscriptionHelpers {
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/SubscriptionHelpers.java]: Calling this class "Subscriptions" is consistent with other names we've used for factory-style interfaces, e.g. MongoClients.

Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/SubscriptionHelpers.java]: None of these methods are under test, so probably needs a SubscriptionsSpecification of some sort

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client;
+
+/**
+ * A {@link Subscription} represents a one-to-one lifecycle of a {@link Observer} subscribing to a observable.
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: "an observable"

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * <p>
+ * It can only be used once by a single {@link Observer}.
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: "Instances can only be used..."

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 2 above)
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: No need to @link to self

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * </p>
+ * <p>
+ * It is used to both signal desire for data and to allow for unsubscribing.
+ * </p>
+ *
+ * @since 3.1
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: @see SubscriptionHelpers

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ */
+public interface Subscription {
+
+    /**
+     * No operation will be sent to MongoDB from the observable until demand is signaled via this method.
+     * <p>
+     * It can be called however often and whenever needed—but the outstanding cumulative demand must never exceed Long.MAX_VALUE.
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: replace hyphen with either comma or double hyphen surrounded by spaces.

Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Surround Long.MAX_VALUE with @code or @link

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+     * An outstanding cumulative demand of Long.MAX_VALUE may be treated by the observable as "effectively unbounded".
+     * </p>
+     * <p>
+     * Whatever has been requested might be sent, so only signal demand for what can be safely handled.
+     * <p>
+     * A observable can send less than is requested if the stream ends but then must emit either {@link Observer#onError(Throwable)} or
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: "An observable"

File: driver-async/src/main/com/mongodb/async/client/Subscription.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+     * {@link Observer#onComplete()}.
+     * </p>
+     * @param n the strictly positive number of elements to requests to the upstream observable
+     */
+    void request(long n);
+
+    /**
+     * Request the observable to stop sending data and clean up resources.
+     * <p>
+     * Data may still be sent to meet previously signalled demand after calling cancel as this request is asynchronous.
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Move "as this request is asynchronous" to the start of the sentence.

File: driver-async/src/main/com/mongodb/async/client/FlatteningSingleResultCallbackSubscription.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        this.block = block;
+    }
+
+    @Override
+    void requestInitialData() {
+        block.apply(new SingleResultCallback<List<TResult>>() {
+            @Override
+            public void onResult(final List<TResult> result, final Throwable t) {
+                if (t != null) {
+                    onError(t);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/FlatteningSingleResultCallbackSubscription.java]: Can we add a test for this line?

File: driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Override
+    void requestInitialData() {
+        mongoIterable.batchSize(getBatchSize());
+        mongoIterable.batchCursor(new SingleResultCallback<AsyncBatchCursor<TResult>>() {
+            @Override
+            public void onResult(final AsyncBatchCursor<TResult> result, final Throwable t) {
+                if (t != null) {
+                    onError(t);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java]: Can we add a test for this line?

File: driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+                    synchronized (MongoIterableSubscription.this) {
+                        isReading = false;
+                        if (t == null && result == null) {
+                            completed = true;
+                        }
+                    }
+
+                    if (t != null) {
+                        onError(t);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java]: and perhaps this one too

File: driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        this.block = block;
+    }
+
+    @Override
+    void requestInitialData() {
+        block.apply(new SingleResultCallback<TResult>() {
+            @Override
+            public void onResult(final TResult result, final Throwable t) {
+                if (t != null) {
+                    onError(t);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java]: Can we add a test for this condition?

**Comment by jyemin** [comment]
I like how how it DRY'ed up.  Except for Javadoc, the only substantive change I would prefer is a change from SubscriptionHelpers to Subscriptions.

**Comment by jyemin** [comment]
Oh, and since there's no separate ticket for it, don't forget to add some reference documentation for this.

File: driver-async/src/main/com/mongodb/async/client/SubscriptionHelpers.java (hunk 1 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/SubscriptionHelpers.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 2 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 3 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 4 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 5 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 6 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 7 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: Done

File: driver-async/src/main/com/mongodb/async/client/FlatteningSingleResultCallbackSubscription.java (hunk 8 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/FlatteningSingleResultCallbackSubscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java (hunk 9 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java (hunk 10 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoIterableSubscription.java]: Done.

File: driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java (hunk 11 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java]: Done.

**Comment by rozza** [comment]
Took longer than expected, but whilst writing the docs I noticed that it simply didn't make sense not to have an `Observable` class.  Ultimately, it will provide more flexibility as potentially helpers could be written to chain observable processes together eg. 

```
run(updateObservable1, updateObservable2).andThen(findUpdatedDataObservable).subscribe(...)
```

It turns out it simplifies the logic required upstream - as libraries just have to write a single `Observable` adaptor - so I forced pushed an update to rozza/mongo-java-driver-reactivestreams#2

File: driver-async/src/main/com/mongodb/async/client/Subscription.java (hunk 4 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/Subscription.java]: And removed as `Subscriptions` (nee `SubscriptionHelpers`) became obsolete thanks to `Observables`.

**Comment by jyemin** [comment]
This looks fantastic.  Just a few comments, mostly about the docs.

**Comment by rozza** [comment]
Updated rebased and merged

**Pull request closed without merging**